sdk/lib/crt/except/amd64/ehandler.c

```diff
@@ -1,19 +1,143 @@
+
+    /* Get the image base */
+    ImageBase = (ULONG64)DispatcherContext->ImageBase;
+
+    /* Get the image base relative instruction pointers */
+    IpOffset = DispatcherContext->ControlPc - ImageBase;
+    TargetIpOffset = DispatcherContext->TargetIp - ImageBase;
+
+    /* Get the scope table and current index */
+    ScopeTable = (PSCOPE_TABLE)DispatcherContext->HandlerData;
+
+    /* Loop while we have scope table entries */
+    while (DispatcherContext->ScopeIndex < ScopeTable->Count)
+    {
+        /* Use i as index and update the dispatcher context */
+        i = DispatcherContext->ScopeIndex++;
+
+        /* Get the start and end of the scrope */
+        BeginAddress = ScopeTable->ScopeRecord[i].BeginAddress;
+        EndAddress = ScopeTable->ScopeRecord[i].EndAddress;
+
+        /* Skip this scope if we are not within the bounds */
+        if ((IpOffset < BeginAddress) || (IpOffset >= EndAddress))
+        {
+            continue;
+        }
+
+        /* Check if this is an unwind */
+        if (ExceptionRecord->ExceptionFlags & EXCEPTION_UNWIND)
+        {
+            /* Check if this is a target unwind */
+            if (ExceptionRecord->ExceptionFlags & EXCEPTION_TARGET_UNWIND)
+            {
+                /* Check if the target is within the scope itself */
+                if ((TargetIpOffset >= BeginAddress) &&
+                    (TargetIpOffset <  EndAddress))
+                {
+                    return ExceptionContinueSearch;
+                }
+            }
+
+            /* Check if this is a termination handler / finally function */
+            if (ScopeTable->ScopeRecord[i].JumpTarget == 0)
+            {
+                /* Call the handler */
+                Handler = ScopeTable->ScopeRecord[i].HandlerAddress;
+                TerminationHandler = (PTERMINATION_HANDLER)(ImageBase + Handler);
+                TerminationHandler(TRUE, EstablisherFrame);
+            }
+            else if (ScopeTable->ScopeRecord[i].JumpTarget == TargetIpOffset)
+            {
+                return ExceptionContinueSearch;
+            }
+        }
+        else
+        {
+            /* We are only unterested in exception handlers */
+            if (ScopeTable->ScopeRecord[i].JumpTarget == 0)
+            {
+                continue;
+            }
+
+            /* This is an exception filter, get the handler address */
+            Handler = ScopeTable->ScopeRecord[i].HandlerAddress;
+
+            /* Check for hardcoded EXCEPTION_EXECUTE_HANDLER */
+            if (Handler == EXCEPTION_EXECUTE_HANDLER)
+            {
+                /* This is our result */
+                FilterResult = EXCEPTION_EXECUTE_HANDLER;
+            }
+            else
+            {
+                /* Otherwise we need to call the handler */
+                ExceptionFilter = (PEXCEPTION_FILTER)(ImageBase + Handler);
+                FilterResult = ExceptionFilter(&ExceptionPointers, EstablisherFrame);
```

Comment:
Are we sure that `ExceptionFilter` will never ever be NULL? (could be NULL if the exception record gets corrupted)

Reply:
We don't deal with that.